File: server/agent/src/agents/core/AgentBase.js
/**
 * AgentBase — Base Class for All Multi-Agent System Agents
 * 
 * Every agent (Supervisor, Itinerary, Recommender, etc.) extends this class.
 * Provides:
 *   - Identity (name, role, systemPrompt)
 *   - Tool execution via ToolRegistry
 *   - ReAct loop (think → tool-call → observe → think → ... → final answer)
 *   - Per-agent memory (conversation history)
 *   - Inter-agent communication via MessageBus
 */

const llm = require('../../llm/connector');
const messageBus = require('./MessageBus');
const toolRegistry = require('./ToolRegistry');

class AgentBase {
    /**
     * @param {Object} config
     * @param {string} config.name — unique agent name (e.g., 'ItineraryAgent')
     * @param {string} config.role — human-readable role description
     * @param {string} config.systemPrompt — the agent's system prompt
     * @param {string[]} config.toolNames — names of tools this agent can use
     * @param {number} config.maxIterations — max ReAct loop iterations (default: 5)
     * @param {number} config.memoryLimit — max conversation turns to keep (default: 20)
     */
    constructor({ name, role, systemPrompt, toolNames = [], maxIterations = 5, memoryLimit = 20 }) {
        this.name = name;
        this.role = role;
        this.systemPrompt = systemPrompt;
        this.toolNames = toolNames;
        this.maxIterations = maxIterations;
        this.memoryLimit = memoryLimit;
        this.memory = [];
        this.trace = []; // Execution trace for this run
        this.sessionHistory = []; // Injected from API for multi-turn context

        // Register with the message bus
        messageBus.subscribe(this.name, this._handleMessage.bind(this));
    }

    /**
     * Main entry point — process a message through the ReAct loop.
     * Loops: think → tool-call → observe → think → ... → final text answer.
     * 
     * @param {string} message — the user/delegated message
     * @param {Object} context — additional context (sessionId, language, etc.)
     * @returns {Promise<Object>} — { success, result, trace }
     */
    async run(message, context = {}) {
        this.trace = [];
        const startTime = Date.now();

        console.log(`\n[${this.name}] ═══ Starting run ═══`);
        console.log(`[${this.name}] Message: ${message.substring(0, 150)}`);

        // Get tool declarations for this agent
        const toolDeclarations = toolRegistry.getDeclarations(this.toolNames);

        let currentMessage = message;
        let iteration = 0;
        let finalResult = null;
        let toolContext = ''; // Accumulate tool results instead of broken history

        while (iteration < this.maxIterations) {
            iteration++;
            console.log(`[${this.name}] ── Iteration ${iteration}/${this.maxIterations} ──`);

            // Build the message with accumulated tool context
            const fullMessage = toolContext
                ? `Original task: ${message}\n\nTool results so far:\n${toolContext}\n\nBased on the tool results above, either call another tool or provide your final answer.`
                : currentMessage;

            // THINK: Send message + tools to LLM with conversation memory
            const recentMemory = [
                ...this.sessionHistory.slice(-6), // Last 3 exchanges from session DB
                ...this.memory.slice(-6),          // Last 3 exchanges from agent memory
            ].slice(-8); // Cap at 8 total messages for context window
            const thinkResult = await this._think(fullMessage, toolDeclarations, recentMemory, context);

            if (!thinkResult.success) {
                console.error(`[${this.name}] Think failed:`, thinkResult.error);
                this.trace.push({ step: 'error', error: thinkResult.error, iteration });
                finalResult = { success: false, error: thinkResult.error };
                break;
            }

            // Check if LLM returned a function call or a text response
            if (thinkResult.type === 'text') {
                // LLM produced a final text answer — we're done
                console.log(`[${this.name}] Final answer produced (${thinkResult.text.length} chars)`);
                this.trace.push({ step: 'final_answer', iteration, preview: thinkResult.text.substring(0, 200) });
                finalResult = { success: true, text: thinkResult.text, source: thinkResult.source };
                break;
            }

            if (thinkResult.type === 'functionCall') {
                const { name: toolName, args } = thinkResult.functionCall;
                console.log(`[${this.name}] Tool call: ${toolName}(${JSON.stringify(args).substring(0, 100)})`);
                this.trace.push({ step: 'tool_call', tool: toolName, args, iteration });

                // EXECUTE: Run the tool
                const toolResult = await this._executeTool(toolName, args);
                this.trace.push({ step: 'tool_result', tool: toolName, resultPreview: JSON.stringify(toolResult).substring(0, 200), iteration });

                // OBSERVE: Accumulate tool result for next iteration
                const toolResultStr = typeof toolResult === 'string'
                    ? toolResult
                    : JSON.stringify(toolResult);

                toolContext += `\n[Tool: ${toolName}] Args: ${JSON.stringify(args)}\nResult: ${toolResultStr.substring(0, 2000)}\n`;
            }
        }

        if (!finalResult) {
            console.warn(`[${this.name}] Max iterations reached without final answer`);
            finalResult = { success: false, error: 'Max iterations reached' };
        }

        // Add to agent memory
        this.addMemory('user', message);
        if (finalResult.text) {
            this.addMemory('assistant', finalResult.text);
        }

        const elapsed = Date.now() - startTime;
        console.log(`[${this.name}] ═══ Run complete (${elapsed}ms, ${iteration} iterations) ═══\n`);

        return {
            ...finalResult,
            agentName: this.name,
            iterations: iteration,
            trace: this.trace,
            elapsedMs: elapsed
        };
    }

    /**
     * Send a message + tools to the LLM and get back either a tool call or text.
     * @private
     */
    async _think(message, toolDeclarations, history, context) {
        const langInstruction = context.language === 'hi'
            ? '\n\nIMPORTANT: Respond ENTIRELY in Hindi (Devanagari script). Place names stay in English.'
            : '';

        const fullPrompt = this.systemPrompt + langInstruction;

        // If we have tools, use function calling
        if (toolDeclarations.length > 0) {
            const result = await llm.generateWithTools(fullPrompt, message, toolDeclarations, history);
            return result;
        }

        // No tools — standard text generation
        const result = await llm.generateResponse(fullPrompt, message, history);
        if (!result.success) {
            return { success: false, error: result.error };
        }
        return { success: true, type: 'text', text: result.text, source: result.source };
    }

    /**
     * Execute a tool via the ToolRegistry.
     * @private
     */
    async _executeTool(toolName, args) {
        try {
            return await toolRegistry.executeTool(toolName, args);
        } catch (err) {
            console.error(`[${this.name}] Tool execution error (${toolName}):`, err.message);
            return { error: err.message };
        }
    }

    /**
     * Send a message to another agent via the MessageBus.
     * @param {string} toAgent — target agent name
     * @param {string} type — message type
     * @param {*} content — message payload
     * @returns {Promise<*>} response from the target agent
     */
    async sendMessage(toAgent, type, content) {
        return messageBus.request(this.name, toAgent, type, content);
    }

    /**
     * Handle an incoming message from the MessageBus.
     * Subclasses can override this for custom inter-agent behavior.
     * Default: runs the message content through the agent's ReAct loop.
     * @private
     */
    async _handleMessage(message) {
        console.log(`[${this.name}] Received message from ${message.from}: ${message.type}`);
        const content = typeof message.content === 'string'
            ? message.content
            : JSON.stringify(message.content);
        const result = await this.run(content, {});
        return result;
    }

    /**
     * Add a message to the agent's memory (capped at memoryLimit).
     */
    addMemory(role, content) {
        this.memory.push({ role, content });
        if (this.memory.length > this.memoryLimit) {
            this.memory = this.memory.slice(-this.memoryLimit);
        }
    }

    /**
     * Clear the agent's memory.
     */
    clearMemory() {
        this.memory = [];
        this.sessionHistory = [];
    }

    /**
     * Inject session history from the database for multi-turn context.
     * Call this before run() for follow-up messages.
     * @param {Array} history — [{role, content}, ...]
     */
    setSessionHistory(history = []) {
        this.sessionHistory = history;
    }

    /**
     * Get the agent's execution trace.
     */
    getTrace() {
        return [...this.trace];
    }
}

module.exports = AgentBase;

File: server_py/agent/agents/core/agent_base.py
"""
AgentBase — Base class for all BrajYatra agents.
Implements the ReAct (Reasoning + Acting) loop:
  1. Think — Send context to LLM
  2. Act — Execute tool if LLM requests one
  3. Observe — Feed tool result back to LLM
  4. Repeat until LLM produces a final text answer
"""

import json
import time
from typing import Optional

from agent.llm import connector as llm
from agent.agents.core.tool_registry import get_declarations, execute_tool
from agent.agents.core.message_bus import message_bus


class AgentBase:
    """Base agent implementing the ReAct loop."""

    def __init__(
        self,
        name: str,
        system_prompt: str,
        tool_names: list[str] = None,
        max_iterations: int = 5,
        memory_limit: int = 10,
    ):
        self.name = name
        self.system_prompt = system_prompt
        self.tool_names = tool_names or []
        self.max_iterations = max_iterations
        self.memory_limit = memory_limit

        self.memory: list[dict] = []
        self.session_history: list[dict] = []

        # Register on message bus
        message_bus.subscribe(f"request:{name}", self._handle_bus_request)
        print(f"[{self.name}] Initialized with tools: {self.tool_names}")

    def set_session_history(self, history: list[dict]):
        """Inject session history for multi-turn context."""
        self.session_history = history[-6:]  # Keep last 6 messages

    async def run(self, user_message: str, context: dict = None) -> dict:
        """Run the ReAct loop.

        Args:
            user_message: User's input
            context: Additional context (sessionId, language, etc.)

        Returns:
            { success, text?, source?, ... }
        """
        context = context or {}
        start_time = time.time()

        # Get tool declarations for this agent
        tool_decls = get_declarations(self.name)

        # Build the conversation history for the LLM
        history = list(self.session_history)

        # Add agent memory (use "model" role for Gemini compatibility)
        for mem in self.memory[-self.memory_limit:]:
            history.append({"role": "user", "content": mem.get("input", "")})
            history.append({"role": "model", "content": mem.get("output", "")})

        # ReAct loop
        iteration = 0
        accumulated_context = ""

        while iteration < self.max_iterations:
            iteration += 1

            # Build the message with accumulated tool results
            message_for_llm = user_message
            if accumulated_context:
                message_for_llm = f"{user_message}\n\n--- Tool Results ---\n{accumulated_context}"

            # Call LLM
            if tool_decls:
                result = await llm.generate_with_tools(
                    self.system_prompt,
                    message_for_llm,
                    tool_declarations=tool_decls,
                    history=history,
                )
            else:
                result = await llm.generate_response(
                    self.system_prompt,
                    message_for_llm,
                    history=history,
                )
                if result.get("success"):
                    result["type"] = "text"

            if not result.get("success"):
                print(f"[{self.name}] LLM call failed: {result.get('error')}")
                return {
                    "success": False,
                    "text": "I'm having trouble processing your request. Please try again.",
                    "source": self.name,
                    "error": result.get("error"),
                }

            # Check if LLM wants to call a tool
            if result.get("type") == "functionCall":
                fc = result["functionCall"]
                tool_name = fc["name"]
                tool_args = fc.get("args", {})

                print(f"[{self.name}] Iteration {iteration}: Calling tool '{tool_name}' with args: {json.dumps(tool_args)[:200]}")

                # Execute the tool
                try:
                    tool_result = await execute_tool(tool_name, tool_args)
                    tool_result_str = json.dumps(tool_result, default=str, ensure_ascii=False)

                    # Accumulate context for next iteration
                    accumulated_context += f"\n[Tool: {tool_name}] Result:\n{tool_result_str}\n"

                    # Add to conversation history for context
                    history.append({"role": "user", "content": message_for_llm})
                    history.append({"role": "model", "content": f"I called {tool_name} and got: {tool_result_str[:500]}"})

                except Exception as e:
                    print(f"[{self.name}] Tool execution error: {e}")
                    accumulated_context += f"\n[Tool: {tool_name}] Error: {str(e)}\n"

                continue  # Go to next iteration

            # LLM returned a final text answer
            text = result.get("text", "")
            elapsed = round(time.time() - start_time, 2)

            # Save to memory
            self.memory.append({
                "input": user_message,
                "output": text,
                "timestamp": time.time(),
            })
            if len(self.memory) > self.memory_limit:
                self.memory = self.memory[-self.memory_limit:]

            print(f"[{self.name}] Completed in {elapsed}s ({iteration} iterations)")

            return {
                "success": True,
                "text": text,
                "source": self.name,
                "iterations": iteration,
                "elapsed": elapsed,
                **({} if not result.get("groundingMetadata") else {"groundingMetadata": result["groundingMetadata"]}),
            }

        # Max iterations reached
        print(f"[{self.name}] Max iterations ({self.max_iterations}) reached")
        return {
            "success": True,
            "text": accumulated_context if accumulated_context else "I ran out of steps while processing your request.",
            "source": self.name,
            "iterations": iteration,
            "warning": "max_iterations_reached",
        }

    async def _handle_bus_request(self, data: dict) -> dict:
        """Handle a request from the MessageBus (inter-agent delegation)."""
        task = data.get("task", "") if isinstance(data, dict) else str(data)
        context = data.get("context", {}) if isinstance(data, dict) else {}
        return await self.run(task, context)
